### PR TITLE
feat: add `reth db snapshot <TYPE>` command

### PR DESCRIPTION
**Pull request by joshieDo**
* adds `reth db snapshot` command to **generate snapshots** from different categories, different configurations in one command, **alongside their benchmarks**.
    * allows for better development, debugging and experimentation
    * allows users to generate their own snapshots  and just plug them in.  Maybe they want it highly compressed in a certain range, but uncompressed in another specific range.
* adds some `RUST_LOG=debug` logs
* adds `lz4` & query improvements ( stacked from https://github.com/paradigmxyz/reth/pull/4890 )

If the approach is accepted, I'll be adding the other  categories (`transactions`, `receipts`, `blocks`, etc).

Regarding benchmarks I'm re-initializing the database environment to try and make sure we query it cold.

### CODE REVIEW

File: bin/reth/src/db/snapshots/mod.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    /// Compression algorithms to use.
+    #[arg(long, short, value_delimiter = ',', default_value = "lz4")]
+    compression: Vec<Compression>,
+
+    /// Flag to enable inclusion list filters and PHFs.
+    #[arg(long, default_value = "true")]
+    with_filters: bool,
+
+    /// Specifies the perfect hashing function to use.
+    #[arg(long, value_delimiter = ',', default_value_if("with_filters", "true", "mphf"))]
```
Comment (mattsse) [inline bin/reth/src/db/snapshots/mod.rs:52]: TIL

File: bin/reth/src/db/snapshots/mod.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            "snapshot_{mode:?}_{}_{}_{compression:?}_{phf:?}",
+            self.from,
+            self.from + self.block_interval
+        )
+        .into()
+    }
+}
+
+pub(crate) type Rows = Vec<Vec<Vec<u8>>>;
+pub(crate) type JarConfig = (Snapshots, Compression, PerfectHashingFunction);
```
Comment (mattsse) [inline bin/reth/src/db/snapshots/mod.rs:195]: move to top of file

File: bin/reth/src/db/snapshots/headers.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                    .hash_slow();
+
+            bench(
+                BenchKind::RandomHash,
+                (open_db_read_only(db_path, log_level)?, chain.clone()),
+                jar_config,
+                || {
+                    let header = Header::decompress(
+                        cursor
+                            .row_by_key_with_cols::<0b01, 2>(header_hash.as_slice())?
```
Comment (mattsse) [inline bin/reth/src/db/snapshots/headers.rs:174]: how many variants of this call do we have?

I kinda like the consts here, but if we have _a lot_ of them perhaps we should consider args instead

**Review by mattsse — APPROVED** [review]
love the command!

I did not review the compression changes in detail, on a high level they look good.
test coverage gives me the confidence to approve this though, but would appreciate if @shekhirin would take a look as well.

I see we're using files directly, did we check Buffered IO as well?

**Review by shekhirin — APPROVED** [review]
LGTM, only nits

File: bin/reth/src/db/snapshots/headers.rs (hunk 3 above)
Comment (joshieDo) [reply bin/reth/src/db/snapshots/headers.rs:174]: i'd reckon we wont have many tbh, but ill keep it in mind for the future

**Pull request merged**